File: backend/lib/block_lettering/pen_pressure.m
function results = compute_pen_pressure_consistency(image_path, debug)
    % Set default for debug if not provided
    if nargin < 2
        debug = false;
    end

    % Read and preprocess the image
    img = imread(image_path);
    if isempty(img)
        disp(['Error: Could not read image at ', image_path]);
        results = struct();
        return;
    end

    % Convert to grayscale if image is RGB
    if size(img, 3) == 3
        gray = rgb2gray(img);
    else
        gray = img;
    end

    % Apply binary thresholding and invert
    threshold = 127 / 255;
    binary = imbinarize(gray, threshold);
    binary = ~binary;  % Invert to match convention

    % Calculate distance transform
    dist_transform = bwdist(~binary);

    % Get stroke thickness measurements (non-zero values represent thickness)
    thickness_values = dist_transform(binary);

    % Skip processing if no strokes detected
    if isempty(thickness_values)
        results = struct('mean_thickness', 0, ...
                        'thickness_std', 0, ...
                        'thickness_variance', 0, ...
                        'coefficient_of_variation', 0);
        return;
    end

    % Calculate statistical measures
    mean_thickness = mean(thickness_values(:));
    thickness_std = std(thickness_values(:));
    thickness_variance = var(thickness_values(:));
    coefficient_of_variation = thickness_std / mean_thickness;

    % Pack results into a structure
    results = struct(...
        'mean_thickness', mean_thickness, ...
        'thickness_std', thickness_std, ...
        'thickness_variance', thickness_variance, ...
        'coefficient_of_variation', coefficient_of_variation ...
    );

    % Debug visualization if requested
    if debug
        figure('Name', 'Pen Pressure Analysis', 'NumberTitle', 'off');

        % Original Image
        subplot(2,2,1);
        imshow(img);
        title('Original Image');

        % Binary Image
        subplot(2,2,2);
        imshow(binary);
        title('Binary Image');

        % Distance Transform Visualization
        subplot(2,2,3);
        imshow(dist_transform, []);
        colormap(gca, jet);
        colorbar;
        title('Distance Transform (Stroke Thickness)');

        % Histogram of thickness values
        subplot(2,2,4);
        histogram(thickness_values, 30);
        hold on;
        xline(mean_thickness, 'r--', 'LineWidth', 2);
        hold off;
        title(sprintf('Thickness Distribution\nMean: %.2f, CoV: %.2f', ...
              mean_thickness, coefficient_of_variation));

        % Print debug information
        fprintf('Mean stroke thickness: %.3f\n', mean_thickness);
        fprintf('Thickness standard deviation: %.3f\n', thickness_std);
        fprintf('Coefficient of variation: %.3f\n', coefficient_of_variation);
    end
end

% === Example Usage ===
image_path = '/Users/jameswong/PycharmProjects/NoteMercy_Extension/backend/atest/1.png';
pressure_results = compute_pen_pressure_consistency(image_path, true);
disp(pressure_results);

File: backend/lib_py/block_lettering/pen_pressure.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

def compute_pen_pressure_consistency(image_path, debug=False):
    # Read the image
    img = cv2.imread(image_path, cv2.IMREAD_UNCHANGED)
    if img is None:
        print(f"Error: Could not read image at {image_path}")
        return {}

    # Convert to grayscale if image is in color
    if len(img.shape) == 3:
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    else:
        gray = img

    # Apply binary thresholding and invert (threshold value of 127, max value 255)
    # cv2.THRESH_BINARY_INV makes the foreground (pen stroke) white and background black.
    ret, binary = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY_INV)

    # Calculate distance transform
    # This computes, for each non-zero (foreground) pixel, the distance to the nearest zero (background) pixel.
    dist_transform = cv2.distanceTransform(binary, cv2.DIST_L2, 5)

    # Get stroke thickness measurements: only consider pixels where binary is non-zero
    thickness_values = dist_transform[binary > 0]

    # If no stroke pixels were detected, return zeros for all measures
    if thickness_values.size == 0:
        return {
            'mean_thickness': 0,
            'thickness_std': 0,
            'thickness_variance': 0,
            'coefficient_of_variation': 0
        }

    # Calculate statistical measures
    mean_thickness = np.mean(thickness_values)
    thickness_std = np.std(thickness_values)
    thickness_variance = np.var(thickness_values)
    coefficient_of_variation = thickness_std / mean_thickness

    results = {
        'mean_thickness': mean_thickness,
        'thickness_std': thickness_std,
        'thickness_variance': thickness_variance,
        'coefficient_of_variation': coefficient_of_variation
    }

    # Debug visualization if requested
    if debug:
        plt.figure("Pen Pressure Analysis", figsize=(10, 8))

        # Original Image
        plt.subplot(2, 2, 1)
        # Convert BGR to RGB for correct display if the image is color
        if len(img.shape) == 3:
            img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
            plt.imshow(img_rgb)
        else:
            plt.imshow(img, cmap='gray')
        plt.title("Original Image")
        plt.axis('off')

        # Binary Image
        plt.subplot(2, 2, 2)
        plt.imshow(binary, cmap='gray')
        plt.title("Binary Image")
        plt.axis('off')

        # Distance Transform Visualization
        plt.subplot(2, 2, 3)
        plt.imshow(dist_transform, cmap='jet')
        plt.colorbar()
        plt.title("Distance Transform (Stroke Thickness)")
        plt.axis('off')

        # Histogram of thickness values
        plt.subplot(2, 2, 4)
        plt.hist(thickness_values, bins=30)
        plt.axvline(mean_thickness, color='r', linestyle='--', linewidth=2)
        plt.title(f"Thickness Distribution\nMean: {mean_thickness:.2f}, CoV: {coefficient_of_variation:.2f}")
        plt.xlabel("Thickness")
        plt.ylabel("Frequency")

        plt.tight_layout()
        plt.show()

        # Print debug information
        print(f"Mean stroke thickness: {mean_thickness:.3f}")
        print(f"Thickness standard deviation: {thickness_std:.3f}")
        print(f"Coefficient of variation: {coefficient_of_variation:.3f}")

    return results

# === Example Usage ===
if __name__ == '__main__':
    image_path = '/Users/jameswong/PycharmProjects/NoteMercy_Extension/backend/atest/1.png'
    pressure_results = compute_pen_pressure_consistency(image_path, debug=True)
    print(pressure_results)
